### ISSUE TITLE
Exception handling design for mspass

### ISSUE THREAD

**Issue by pavlis** [wangyinz/mspass#11]
This emerged from ruminations on this topic by me (Pavlis) followed by the question to Ian about how spark handles a process that calls exit.   His answer was that the entire job terminates - a bad thing for a massive  job.   Hence, we need a way to have a program handle most errors without calling exit.   Part of that will be a coding standard, but for now the topic is the approach to use an API to implement the approach. 

Some things known to not work:

1. Simple error logging to stderr.    The iostreams library is not thread safe and as a result multiple processes writing to stderr will frequently produce jumbled output.
2.  The same happens if logging to a file as in unix file io is no different that output to stderr.
3.  Might consider every process writing it's own log file, but that that could become a bookeeping mess with a confusing problem of how to find the right log file.

I think the clear solution for mspass is to make use of mongodb.   That is, all processes should have an error log connection and the system should automatically save all log messages of varying severity in a mongodb collection.   I think the concept maps easily to having documents in the collection that contain a minimum of these set of attributes:
jobid
processid
objectid of data object being handled that created the log entry
algorithm - the algorithm that was running on the data
severity - describes how bad the problem was.  antelope's elog provides a reasonable list.  They use:  log, notify, debug,complain, die.

We might consider two forms of die:  zombify and abort, but that might create more problems that it solves.  zombify would make a process stop running but wait for other workers to exit.   Not sure that is smart.   Probably cleaner to just reserve die for unrecoverable system problems like a malloc failure.   Worth a serious discussion.  

The secondary thing this solves is the verbose error log problem that plagues systems like antelope.   There have been so many instances of issues that arose because an important program like their dbverify writes so much output it is very difficult to separate the wheat from the chaff.  Putting all error messages as separate documents should solve this problem and create a thread safe log system.   I think we may have mentioned this use of mongo in the proposal, but it is pretty clear to me that is the solution.  Do we agree?

Now a key question is how programs should be designed to handle exceptions in mspass?   I think this has two elements:
1.   The base error message thrown in c++ and python algorithms should have the error severity contained in the exception object for retrieval.   Then error handlers will be able to handle different levels of errors differently.
2.    We need to either design a error logging object that any process running in the system should create or design a procedural set of functions like Antelope's elog.   The later would make it easier for plain C programmers, but likely make the capability more limited.  

The first is trivial, the second is a major task.   I think, however, that if we are clever it could be abstracted as little different from adding new data to the database.   An error message is just another data object.   

Before I try to be more concrete in an API I'd like to get some feedback on this.  Realize it is a lot to digest, but it is a very important issue.

**Comment by pavlis** [comment]
You are right that logging input parameters is very important, but I think it would be better logged to a different collection called something like input_paramters.

I'm starting on the error object part of this.  Ran across an interesting C++ construct in C11 and higher.   
struct DataError
{
	typedef enum{
		Fatal,
		Invalid,
		Suspect,
		Complaint,
		Debug,
		Informational
	} Severity;
};

You can declare one of these as:
DataError::Severity x;

then you can use x in a switch case like this:
switch(x)
{
case Fatal:
case Invalid: 
etc.
};
Useful as otherwise an enum with a readable am can create name collisions very easily.

**Comment by wangyinz** [comment]
Checked by Codex

This looks like a good manual-close candidate. The design discussed here is now represented by core pieces such as `ErrorLogger`, `ErrorSeverity`, and `MsPASSError`, plus database/error-handling utilities in the Python layer. I left it open but marked it for maintainer review as likely superseded by the current implementation.

**Comment by wangyinz** [comment]
Closing this as completed/superseded by the current implementation.

The design proposed here is now represented by `ErrorLogger`/`LogData`, `ErrorSeverity`, and `MsPASSError`; seismic data and processing-history objects carry an `elog`, the classes are exposed through the Python bindings, and the database/`Undertaker` paths persist and handle those logs. The original design question therefore has an established implementation.

Any remaining concrete logging behavior can be tracked in a focused follow-up issue.
